presto-docs/src/main/sphinx/admin/tuning.rst

```diff
@@ -5,35 +5,548 @@ Tuning Presto
+
+ * **Type:** ``Boolean``
+ * **Default value:** ``true``
+ * **Description:** Serves as default value for ``optimize_hash_generation`` session property. Compute hash codes for distribution, joins, and aggregations early in query plan which may allow to drop some of computation later in query processing with the cost of increased preprocessing. In most cases it should decrease overall query processing time.
+
+
+``optimizer.optimize-metadata-queries``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+ * **Type:** ``Boolean``
+ * **Default value:** ``false``
+ * **Description:** Enables optimization of aggregations that are kept in metadata of data structures. This allow to perform simple queries in ``O(1)`` time using metadata that are kept anyway. Currently this optimization can be use for selecting `max`, `min` and `approx_distinct` of partition keys. Using this may speed some queries significantly with possible drawback on very small data sets.
+
+
+``optimizer.optimize-single-distinct``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+ * **Type:** ``Boolean``
+ * **Default value:** ``true``
+ * **Description:** Enables single distinct optimization. This optimization allows to perform applying distinct mask only once in cases where it's possible. This optimization will try to use single GROUP BY instead of multiple DISTINCT clauses. Enabling this optimization should speed up some specific selects but analyzing all queries to check if they qualify for this optimization may be a slight overhead.
+
+
+``optimizer.push-table-write-through-union``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+ * **Type:** ``Boolean``
+ * **Default value:** ``true``
+ * **Description:** Serves as default value for ``push_table_write_through_union`` session property. Parallelize writes when using UNION ALL in queries that write data. This allows to improve speed of writing output tables in UNION ALL clause by making use of the fact, that UNION ALL outputs do not require additional synchronization when collecting results. Enabling this optimization can improve UNION ALL speed when write speed is not yet saturated. However it's may slow down queries in already heavy loaded system.
+
+
+.. _tuning-pref-query:
+
+Query execution properties
+--------------------------
+
+
+``query.execution-policy``
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+ * **Type:** ``String`` (``all-at-once`` or ``phased``)
+ * **Default value:** ``all-at-once``
+ * **Description:** Serves as default value for ``execution_policy`` session property. Setting this value to ``phased`` will allow query scheduler to split a single query execution between different time slots. This will allow to switch context more often and possibly stage the partially executed query in order to increase robustness. Average time of executing query may slightly increase after setting this to ``phased`` due to context switching and more complex scheduling algorithm but drop in variation of query execution time is expected.
```

Comment:
Setting this value to ``phased`` will allow the query scheduler to split a single query execution between different time slots. This will allow Presto to switch context more often and possibly stage the partially executed query in order to increase robustness. Average time to execute a query may slightly increase after setting this to ``phased``, but query execution time will be more consistent.  This can also be specified on a per-query basis using the ``execution_policy`` session property.